File: 210-course-schedule-ii/leochang/210.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<pair<int, int>>& prerequisites) {
		vector<vector<int>> gragh(numCourses);
        vector<int> ans;
		vector<bool> visited(numCourses, false), path(numCourses, false);

		for (auto pre : prerequisites) {
			gragh[pre.second].push_back(pre.first);
		}

		for (int i = 0; i < numCourses; i++) {
			if (!visited[i] && dfs(i, gragh, visited, path, ans)){
				return {};
			}
		}
    reverse(ans.begin(), ans.end());
		return ans;
    }
private:
	bool dfs(int node, vector<vector<int>> &gragh, vector<bool> &visited, vector<bool> &path, vector<int>& ans) {
		if (visited[node]) {
			return false;
		}

		visited[node] = true;
		path[node] = true;

		for (int neighbor : gragh[node]) {
			if (path[neighbor] || dfs(neighbor, gragh, visited, path, ans)) {
				return true;
			}
		}
        
    ans.push_back(node);
		return path[node] = false;
	}
};

File: 210-course-schedule-ii/ken/210.py
class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """
        import collections

        edge_count = collections.defaultdict(int)
        for i in xrange(numCourses):
            edge_count[i] = 0
        for pre in prerequisites:
            edge_count[pre[0]] += 1

        ans = []
        while True:
            s = -1
            for k, v in edge_count.iteritems():
                if v == 0:
                    s = k
                    break

            if s == -1:
                break

            ans.append(s)

            del edge_count[s]
            for pre in prerequisites:
                if pre[1] == s:
                    edge_count[pre[0]] -= 1

        for k, v in edge_count.iteritems():
            if v > 0:
                return []

        return ans
